File: client_app/lib/features/merchants/data/models/catalog_model.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

part 'catalog_model.g.dart';

@JsonSerializable()
class CatalogModel extends Equatable {
  final String id;
  final String merchantId;
  final String name;
  final String? description;
  final String status;
  final List<CategoryModel>? categories;
  final DateTime createdAt;
  final DateTime updatedAt;

  const CatalogModel({
    required this.id,
    required this.merchantId,
    required this.name,
    this.description,
    required this.status,
    this.categories,
    required this.createdAt,
    required this.updatedAt,
  });

  factory CatalogModel.fromJson(Map<String, dynamic> json) =>
      _$CatalogModelFromJson(json);

  Map<String, dynamic> toJson() => _$CatalogModelToJson(this);

  @override
  List<Object?> get props => [
        id,
        merchantId,
        name,
        description,
        status,
        categories,
        createdAt,
        updatedAt,
      ];
}

@JsonSerializable()
class CategoryModel extends Equatable {
  final String id;
  final String catalogId;
  final String name;
  final String? description;
  final String? imageUrl;
  final bool visible;
  final int sortOrder;
  final List<ProductModel>? products;

  const CategoryModel({
    required this.id,
    required this.catalogId,
    required this.name,
    this.description,
    this.imageUrl,
    required this.visible,
    required this.sortOrder,
    this.products,
  });

  factory CategoryModel.fromJson(Map<String, dynamic> json) =>
      _$CategoryModelFromJson(json);

  Map<String, dynamic> toJson() => _$CategoryModelToJson(this);

  @override
  List<Object?> get props => [
        id,
        catalogId,
        name,
        description,
        imageUrl,
        visible,
        sortOrder,
        products,
      ];
}

@JsonSerializable()
class ProductModel extends Equatable {
  final String id;
  final String merchantId;
  final String categoryId;
  final String name;
  final String? description;
  final List<String>? imageUrls;
  final double price;
  final String currency;
  final String availability;
  final StockInfoModel? stockInfo;
  final List<ProductModifierModel>? modifiers;
  final DateTime createdAt;
  final DateTime updatedAt;

  const ProductModel({
    required this.id,
    required this.merchantId,
    required this.categoryId,
    required this.name,
    this.description,
    this.imageUrls,
    required this.price,
    required this.currency,
    required this.availability,
    this.stockInfo,
    this.modifiers,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ProductModel.fromJson(Map<String, dynamic> json) =>
      _$ProductModelFromJson(json);

  Map<String, dynamic> toJson() => _$ProductModelToJson(this);

  @override
  List<Object?> get props => [
        id,
        merchantId,
        categoryId,
        name,
        description,
        imageUrls,
        price,
        currency,
        availability,
        stockInfo,
        modifiers,
        createdAt,
        updatedAt,
      ];
}

@JsonSerializable()
class StockInfoModel extends Equatable {
  final int? quantity;
  final bool trackStock;
  final int? lowStockThreshold;

  const StockInfoModel({
    this.quantity,
    required this.trackStock,
    this.lowStockThreshold,
  });

  factory StockInfoModel.fromJson(Map<String, dynamic> json) =>
      _$StockInfoModelFromJson(json);

  Map<String, dynamic> toJson() => _$StockInfoModelToJson(this);

  @override
  List<Object?> get props => [quantity, trackStock, lowStockThreshold];
}

@JsonSerializable()
class ProductModifierModel extends Equatable {
  final String id;
  final String name;
  final String type;
  final bool required;
  final List<ModifierOptionModel> options;

  const ProductModifierModel({
    required this.id,
    required this.name,
    required this.type,
    required this.required,
    required this.options,
  });

  factory ProductModifierModel.fromJson(Map<String, dynamic> json) =>
      _$ProductModifierModelFromJson(json);

  Map<String, dynamic> toJson() => _$ProductModifierModelToJson(this);

  @override
  List<Object?> get props => [id, name, type, required, options];
}

@JsonSerializable()
class ModifierOptionModel extends Equatable {
  final String id;
  final String name;
  final double? priceAdjustment;

  const ModifierOptionModel({
    required this.id,
    required this.name,
    this.priceAdjustment,
  });

  factory ModifierOptionModel.fromJson(Map<String, dynamic> json) =>
      _$ModifierOptionModelFromJson(json);

  Map<String, dynamic> toJson() => _$ModifierOptionModelToJson(this);

  @override
  List<Object?> get props => [id, name, priceAdjustment];
}

File: merchant-backoffice/src/app/features/catalog/models/catalog.model.ts
export interface Catalog {
  id: string;
  merchantId: string;
  name: string;
  description: string;
  status: CatalogStatus;
  categories: Category[];
  createdAt: Date;
  updatedAt: Date;
}

export interface Category {
  id: string;
  catalogId: string;
  name: string;
  description: string;
  displayOrder: number;
  products: Product[];
  createdAt: Date;
  updatedAt: Date;
}

export interface Product {
  id: string;
  categoryId: string;
  merchantId: string;
  name: string;
  description: string;
  imageUrls: string[];
  price: number;
  currency: string;
  availability: ProductAvailability;
  stockInfo?: StockInfo;
  modifiers: ProductModifier[];
  createdAt: Date;
  updatedAt: Date;
}

export interface ProductModifier {
  id: string;
  name: string;
  type: ModifierType;
  required: boolean;
  options: ModifierOption[];
}

export interface ModifierOption {
  id: string;
  name: string;
  priceAdjustment: number;
  available: boolean;
}

export interface StockInfo {
  quantity: number;
  lowStockThreshold: number;
  trackStock: boolean;
}

export enum CatalogStatus {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  DRAFT = 'DRAFT'
}

export enum ProductAvailability {
  AVAILABLE = 'AVAILABLE',
  OUT_OF_STOCK = 'OUT_OF_STOCK',
  DISCONTINUED = 'DISCONTINUED'
}

export enum ModifierType {
  SINGLE_SELECT = 'SINGLE_SELECT',
  MULTI_SELECT = 'MULTI_SELECT',
  TEXT_INPUT = 'TEXT_INPUT'
}

// Request/Response DTOs
export interface CreateCatalogRequest {
  name: string;
  description: string;
}

export interface UpdateCatalogRequest {
  name?: string;
  description?: string;
  status?: CatalogStatus;
}

export interface CreateCategoryRequest {
  catalogId: string;
  name: string;
  description: string;
  displayOrder?: number;
}

export interface UpdateCategoryRequest {
  name?: string;
  description?: string;
  displayOrder?: number;
}

export interface CreateProductRequest {
  categoryId: string;
  name: string;
  description: string;
  price: number;
  currency: string;
  imageUrls?: string[];
  stockInfo?: StockInfo;
  modifiers?: ProductModifier[];
}

export interface UpdateProductRequest {
  name?: string;
  description?: string;
  price?: number;
  imageUrls?: string[];
  availability?: ProductAvailability;
  stockInfo?: StockInfo;
  modifiers?: ProductModifier[];
}

export interface BulkProductOperation {
  operation: 'UPDATE_AVAILABILITY' | 'UPDATE_PRICE' | 'DELETE';
  productIds: string[];
  data?: any;
}

export interface ImportProductData {
  name: string;
  description: string;
  categoryName: string;
  price: number;
  currency: string;
  availability: string;
  stockQuantity?: number;
}